feat(starter-filters) make the filters panel collapsible and hidden by default (#894)

eduzavarce opened the pull request:
## Summary
This Pull Request introduces the ability to toggle collapsible filters for challenges, enhancing user interaction and accessibility. It includes changes to the HTML structure, styles, functionality, and associated tests, ensuring a smoother and more intuitive filtering experience.

### Main Changes:
- **Collapsible Filters:**
  - Added collapsible sections for languages with animations and ARIA attributes for better accessibility.
  - Introduced "Select All" functionality for tags within languages.
  - Updated filtering logic to handle new collapsible filter behavior in `starter-filters.component`.

- **UI and Styling Improvements:**
  - Adjusted the layout for better spacing and responsiveness.
  - Modified styles for compact visual elements and refined hover/focus experiences.
  - Redesigned toggle buttons and collapsible panels with proper transitions.

- **Behavior Updates:**
  - Applied progress filtering client-side based on user solution statuses.
  - Updated existing filtering logic to work seamlessly with collapsible components.

- **Tests:**
  - Added tests verifying collapsible sections' default and toggled states.
  - Created tests covering "Select All" functionality and its interaction with existing filters.

- **Localization:**
  - Added a new localized string `"all"` for "Select All" in supported languages (English, Spanish, Catalan).

***
### Screenshots:

* Before:
<img width="368" height="835" alt="image" src="https://github.com/user-attachments/assets/11d888ff-2e5f-431d-a172-4030de84437f" />

* After:
![filtros](https://github.com/user-attachments/assets/92946945-dd2e-41ea-8a4e-6fda974b6821)



***
### Checklist

- [x] Tests for the changes have been added, if applicable.
- [x]  Documentation has been updated, if needed.
- [x] All new and existing tests pass consistently.

eduzavarce committed:
fix: ensure full-height layout for main container and add regression tests

eduzavarce committed:
fix: simplify test assertions in main component specs

eduzavarce committed:
fix: bump version to 3.11.5-RELEASE and document main container scrollbar fix in changelog

eduzavarce committed:
feat: add collapsible filters UI for languages, levels, and progress

eduzavarce committed:
fix: adjust filters UI for improved alignment and responsiveness

- Updated HTML structure for consistent chevron alignment.
- Made layout adjustments to buttons using `inline-flex` for better sizing.
- Reduced chevron size and fine-tuned its rotation behavior.
- Adjusted desktop filters column to shrink-to-content for better responsiveness.

eduzavarce committed:
fix: refine filters UI spacing and remove redundant margin

- Reduced gap between title and chevron for better alignment.
- Removed unnecessary top margin from filters section title.

eduzavarce committed:
feat: implement "All" option for language tags and streamline filters UI

- Added "All" checkbox for selecting/deselecting all tags within a language.
- Updated filters logic to handle "All" selection state and synchronize with individual tags.
- Removed collapsible sections for levels and progress filters.
- Adjusted UI spacing, styles, and compacted checkboxes for a cleaner layout.
- Added translations for "All" in multiple languages.

eduzavarce committed:
feat: enhance progress filters with solution status mapping

- Added client-side filtering for challenges based on solution statuses (Not Started, In Progress, Finished).
- Updated logic to handle dynamic filters and sorting integration.
- Adjusted fetch behavior to ensure statuses are retrieved for admin users.
- Added comprehensive test coverage for progress filtering functionality.

eduzavarce committed:
fix: improve filters logic and clean up unused styles

- Replaced object key-based filtering with simpler `Object.values` iteration.
- Fixed inconsistent test sorting by introducing `localeCompare`.
- Removed redundant `#filters-container` styles for better maintainability.

eduzavarce committed:
fix: add spacing between filters and challenges for better UI layout

Ilmira83 reviewed (approved):
The implemented changes fully meet the task requirements and ensure proper toggle behavior with improved UX. Great and thorough work!

sofrayala commented on src/app/modules/starter/components/starter-filters/starter-filters.component.ts:
```diff
@@ -257,8 +231,87 @@ export class StarterFiltersComponent implements OnInit {
+  areAllTagsSelected(langKey: string): boolean {
+    const group = this.tagsForm.get(langKey) as FormGroup | null;
+    const languagesGroup = this.filtersForm.get('languages') as FormGroup | null;
+    const langCtrl = languagesGroup?.get(langKey);
```
You should add a warning line right after you get langCtrl, because that is the moment you know whether the language control exists or not. 
If languagesGroup is present but the specific language key is missing, it usually means something went wrong when building the form. Adding the warning there helps you catch this mistake early instead of quietly returning false and making the bug harder to find later.

`
if (languagesGroup && !langCtrl) { console.warn(Missing language control for key: ${langKey}); }
`

sofrayala reviewed (changes requested):
Congrats!
It was all very well done, the purpose of the task is completed and fucntions work as expected :)
UX: adding a per-language “All” option makes tag selection clearer and is a common pattern.
Accessibility: using [indeterminate] on the input is good, it communicates partial selection to assistive tech when combined with appropriate labels.
Tests: new unit tests cover the new “All” behavior and the inter-language interaction case (language with no tags).
SCSS: more compact checkboxes + focus ring improves keyboard focus visibility.

eduzavarce committed:
fix: make challenge filters collapsible and update version to 3.11.6-RELEASE

eduzavarce committed:
Merge branch 'develop' into feat/894-make_the_filters_panel_collapsible_and_hidden_by_default

ivilarop commented on CHANGELOG.md:
```diff
@@ -4,6 +4,11 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to
 [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+### [ita-challenges-frontend-3.11.6-RELEASE] - 2025-11-25
```
If the product/PO considers this to be "fixing a usability issue" that was already defined as a bug, then it could remain as Fixed, but it would still be better to upgrade to MINOR (3.12.0) because the default behavior changes and may impact the user.

ivilarop commented:
@eduzavarce 

Great job with the documentation and the level of detail in the summary 👏

For this PR, there's no need to split anything, but here's a recommendation for future submissions:

This PR groups several different improvements (collapsible + logic changes + Select All + UI refactoring + i18n + tests).

In a real-world environment, these would typically be separated into smaller PRs:

1️⃣ Main feature

2️⃣ Refactoring or visual changes

3️⃣ Tests / i18n if applicable

This greatly simplifies the review process, reduces conflicts, and allows for safer rollbacks.

No need to change anything here, just keep this in mind for future submissions.

ivilarop commented on src/app/modules/starter/components/starter/starter.component.ts:
```diff
@@ -65,7 +72,7 @@ export class StarterComponent implements OnInit {
       this.isAdmin = role === 'ADMIN'
       this.cd.detectChanges()
-      if (!this.isAdmin) {
+      if (role && role !== '') {
```
Why this change?

eduzavarce replied:
Already updated this, changed my mistake and it was a requirement in user story 542. thanks for pointing it out!

ivilarop commented on src/app/modules/starter/components/starter-filters/starter-filters.component.ts:
```diff
@@ -27,6 +27,11 @@ export class StarterFiltersComponent implements OnInit {
   public isUserLoggedIn: boolean = false
 
+  // Collapsible state: all groups collapsed by default
+  public openLanguages: Record<string, boolean> = {}
```
From a functional standpoint, it's fine to have a state for each group (languages, levels, progress).

If you wanted to refine it in the future, we could unify it into a common structure (for example, a panel map) to avoid having several separate properties (openLevels, openProgress, etc.).

But this isn't blocking the Homebrew experience; the current approach is correct, although a bit more verbose.

ivilarop commented on src/app/modules/starter/components/starter/starter.component.ts:
```diff
@@ -147,14 +154,35 @@ export class StarterComponent implements OnInit {
+
+      // Apply progress filters client-side based on user solution statuses
+      if (this.filters.progress && this.filters.progress.length > 0) {
+        const wantsNotStarted = this.filters.progress.includes(1);
```
Avoid the magic numbers 1, 2, 3. Can we use enums?

eduzavarce replied:
created an enum to fix it accordingly. thanks!

eduzavarce committed:
fix: update progress filters to use `SolutionStatus` enum and optimize admin logic

- Replaced hardcoded progress filter values with `SolutionStatus` enum for better maintainability.
- Refined admin logic to skip unnecessary `fetchUserSolutionsStatus` call when user is admin.

eduzavarce committed:
bump version to 3.12.0-RELEASE and update changelog

eduzavarce committed:
refactor: replace `SolutionStatus` with `ProgressFilter` enum in progress filters logic

- Introduced new `ProgressFilter` enum for improved clarity and maintainability.
- Updated filtering logic and test cases to use `ProgressFilter` instead of `SolutionStatus`.

eduzavarce committed:
fix: update test to avoid calling `fetchUserSolutionsStatus` on init for admin users

eduzavarce committed:
refactor: remove `ProgressFilter` enum and update progress logic to use `SolutionStatus`

- Eliminated `ProgressFilter` enum in favor of `SolutionStatus` for consistency.
- Updated filtering logic and test cases to reflect the change.
